compact_str/src/lib.rs

```diff
@@ -686,3 +681,44 @@ impl fmt::Write for CompactStr {
 }
 
 crate::asserts::assert_size_eq!(CompactStr, String);
+
+/// A trait for converting a value to a `CompactStr`.
+///
+/// This trait is automatically implemented for any type which implements the
+/// [`Display`] trait. As such, `ToCompactStr` shouldn't be implemented directly:
+/// [`Display`] should be implemented instead, and you get the `ToCompactStr`
+/// implementation for free.
+///
+/// [`Display`]: fmt::Display
+pub trait ToCompactStr {
+    /// Converts the given value to a `CompactStr`.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// let i = 5;
+    /// let five = compact_str::CompactStr::from("5");
+    ///
+    /// assert_eq!(five, i.to_string());
+    /// ```
+    fn to_compact_str(&self) -> CompactStr;
+}
+
+/// # Panics
+///
+/// In this implementation, the `to_compact_str` method panics
+/// if the `Display` implementation returns an error.
+/// This indicates an incorrect `Display` implementation
+/// since `std::fmt::Write for String` never returns an error itself and
+/// the implementation of `ToCompactStr::to_compact_str` only panic if the `Display`
+/// implementation is incorrect..
+impl<T: fmt::Display + ?Sized> ToCompactStr for T {
+    #[inline]
+    fn to_compact_str(&self) -> CompactStr {
+        CompactStr {
+            repr: Repr::from_fmt(self),
+        }
```

Comment:
`CompactStr` now implements `fmt::Write`, so I think this entire impl could just be
```
fn to_compact_str(&self) -> CompactStr {
  let mut compact = CompactStr::new("");
  write!(compact, "{}", self).expect("fmt::Display incorrectly implemented!");
  compact
}
```

Reply:
That would make the code much shorter.

However, I think we should benchmark the code first, as the current implementation might provide better performance than using fmt::Write, since everything is written in one-go.

For size <= MAX_SIZE, current implementation can be optimized into one copy of array, or no copy at all.

For size > MAX_SIZE, current implementation would only allocate once, where fmt::Write could allocate multiple times.

Reply:
I have pushed a new commit that optimize the case where size > `MAX_SIZE` and preallocate the `String` with the exact size and use `fmt::Write` to write the format string.

This would avoid the repeated allocation that occurs if using `fmt::Write` implementation of `CompactStr`.

Reply:
I ran some benchmarks with this implementation and with just using `write!(...)`. In the simple case (e.g. `&str`, `usize`, etc.) just using `write!(...)` is a lot faster, especially for short strings. It's when we have long strings with multiple arguments (e.g. a struct) that counting all of the bytes beforehand and pre-allocating saves a lot of time.

Given this, the approach of initially counting bytes works for me, with the idea that we can optimize for the simple cases later on.

Reply:
I ran some benchmarks with this implementation and with just using `write!(...)`. In the simple case (e.g. `&str`, `usize`, etc.) just using `write!(...)` is a lot faster, especially for short strings. It's when we have long strings with multiple arguments (e.g. a struct) that counting all of the bytes beforehand and pre-allocating saves a lot of time.

Given this, the approach of initially counting bytes works for me, with the idea that we can optimize for the simple cases later on.